agent/src/main/java/com/code_intelligence/jazzer/instrumentor/Hooks.kt

```diff
@@ -0,0 +1,110 @@
+// Copyright 2021 Code Intelligence GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.code_intelligence.jazzer.instrumentor
+
+import com.code_intelligence.jazzer.api.MethodHook
+import com.code_intelligence.jazzer.api.MethodHooks
+import com.code_intelligence.jazzer.utils.ClassNameGlobber
+import io.github.classgraph.ClassGraph
+import io.github.classgraph.ScanResult
+import java.lang.reflect.Method
+
+data class Hooks(
+    val hooks: List<Hook>,
+    val hookClasses: Set<Class<*>>,
+    val additionalHookClassNameGlobber: ClassNameGlobber
+) {
+
+    companion object {
+        fun loadHooks(vararg hookClassNames: Set<String>): List<Hooks> {
+            return ClassGraph()
+                .enableClassInfo()
+                .enableSystemJarsAndModules()
+                .rejectPackages("jaz.*", "com.code_intelligence.jazzer.*")
+                .scan()
+                .use { scanResult ->
+                    // Capture scanResult in HooksLoader field to not pass it through
+                    // all internal hook loading methods.
+                    val loader = HooksLoader(scanResult)
+                    hookClassNames.map(loader::load)
+                }
+        }
+
+        private class HooksLoader(private val scanResult: ScanResult) {
+            fun load(hookClassNames: Set<String>): Hooks {
+                val hooksWithHookClasses = hookClassNames.flatMap(::loadHooks)
+                val hooks = hooksWithHookClasses.map { it.first }
+                val hookClasses = hooksWithHookClasses.map { it.second }.toSet()
+                val additionalHookClassNameGlobber = ClassNameGlobber(
+                    hooks.flatMap(Hook::additionalClassesToHook),
+                    emptyList()
+                )
+                return Hooks(hooks, hookClasses, additionalHookClassNameGlobber)
+            }
+
+            private fun loadHooks(hookClassName: String): List<Pair<Hook, Class<*>>> {
+                return try {
+                    // Kotlin interferes with the caller sensitive Class.forName and no
```

Comment:
I assumed Kotlin did some class loading trickery here because some "agent classes" were loaded by an application class loader (e.g. `AgentJarFinder`) and others by the bootstrap class loader.  `BootstrapClassLoader` is not able to access classes on the classpath, so custom hooks can not be found. This has nothing to do with Kotlin, correct. I updated the comment accordingly.  Thx for questioning this!
Still, I wonder why and where `Agent`, `AgentJarFinder` and some shaded Kotlin classes are loaded.

Reply:
I think that static dependencies of the `Agent` are loaded by an application class loader since that happens before I got a chance to append the agent JAR to the `BootstrapClassLoader`.